### PR TITLE
#2289 - Implement logic for Application Event Code in IER12 - code and date assignment and unit tests (Part 4)

### PR DESCRIPTION
**Pull request by ann-aot**
**Final PR for the story**
- Replaced `appendOptionalStringWithFiller` with `appendStringWithFiller` for both application event code and application event date.
- Created below unit test,
image
image
image
image
image

### CODE REVIEW

File: sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/utils-service/_tests_/unit/application-event-code-during-assessment-utils.service.applicationEventCodeDuringAssessment.spec.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
+import { ApplicationEventCode } from "../../../models/ier12-integration.model";
+import { ApplicationEventCodeDuringAssessmentUtilsService } from "../..";
+import { ApplicationSharedService } from "@sims/services";
+import { createMock } from "@golevelup/ts-jest";
+
+describe("ApplicationEventCodeDuringAssessmentUtilsService-applicationEventCodeDuringAssessment", () => {
+  let applicationEventCodeDuringAssessmentUtilsService: ApplicationEventCodeDuringAssessmentUtilsService;
+  let applicationSharedService: ApplicationSharedService;
+  beforeAll(() => {
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/utils-service/_tests_/unit/application-event-code-during-assessment-utils.service.applicationEventCodeDuringAssessment.spec.ts:9]: We can keep on blank line above `beforeAll` to be aligned with others.

File: sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/utils-service/_tests_/unit/application-event-code-during-enrolment-and-completed-utils.service.applicationEventCodeDuringCompleted.spec.ts [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+      disbursementScheduleStatus: DisbursementScheduleStatus.Pending,
+    };
+  });
+
+  beforeEach(() => {
+    jest.resetAllMocks();
+  });
+
+  it(`Should return ${ApplicationEventCode.DISC} when the disbursement schedule status is ${DisbursementScheduleStatus.Cancelled}.`, async () => {
+    // Arrange and act.
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/utils-service/_tests_/unit/application-event-code-during-enrolment-and-completed-utils.service.applicationEventCodeDuringCompleted.spec.ts:41]: As a suggestion, as much as possible I would like to keep the Arrange/Act/Assert section isolated, as below.
```ts
// Arrange
const currentDisbursementSchedule = {
  ...payload,
  disbursementScheduleStatus: DisbursementScheduleStatus.Cancelled,
};

// Act
const applicationEventCode =
  await applicationEventCodeDuringEnrolmentAndCompletedUtilsService.applicationEventCodeDuringCompleted(
    currentDisbursementSchedule,
  );

// Assert.
expect(applicationEventCode).toBe(ApplicationEventCode.DISC);
```

File: sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/utils-service/_tests_/unit/application-event-code-during-enrolment-and-completed-utils.service.applicationEventCodeDuringCompleted.spec.ts [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+        );
+      // Assert.
+      expect(applicationEventCode).toBe(ApplicationEventCode.COEA);
+    },
+  );
+
+  it(
+    `Should return ${ApplicationEventCode.DISR} when the disbursement schedule status is ${DisbursementScheduleStatus.Pending}, ` +
+      `COE status is ${COEStatus.completed}, disbursement date is same or after the cutoff date(i.e disbursement date + ${DISBURSEMENT_FILE_GENERATION_ANTICIPATION_DAYS}) ` +
+      `and has atleast one ${RestrictionActionType.StopFullTimeDisbursement}.`,
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/utils-service/_tests_/unit/application-event-code-during-enrolment-and-completed-utils.service.applicationEventCodeDuringCompleted.spec.ts:79]: Typo on `atleast`.

File: sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/utils-service/_tests_/unit/application-event-code-during-enrolment-and-completed-utils.service.applicationEventCodeDuringCompleted.spec.ts [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+          } as DisbursementScheduleForApplicationEventCodeDuringCompleted,
+        );
+      // Assert.
+      expect(applicationEventCode).toBe(ApplicationEventCode.DISE);
+    },
+  );
+
+  it(
+    `Should return ${ApplicationEventCode.DISW} when the disbursement schedule status is ${DisbursementScheduleStatus.Sent},` +
+      ` does not have any full time disbursement feedback errors and an award was withheld due to some restriction.`,
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/utils-service/_tests_/unit/application-event-code-during-enrolment-and-completed-utils.service.applicationEventCodeDuringCompleted.spec.ts:166]: Minor but we can have the white spaces all added to the end of the previous line.

File: sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/utils-service/_tests_/unit/application-event-code-during-enrolment-and-completed-utils.service.applicationEventCodeDuringEnrolmentAndCompleted.spec.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    expect(applicationEventCode).toBe(ApplicationEventCode.COED);
+  });
+
+  it(`Should throw an error when the coe status is other than ${COEStatus.required} or ${COEStatus.declined}.`, () => {
+    // Arrange, act and assert.
+    expect(() => {
+      applicationEventCodeDuringEnrolmentAndCompletedUtilsService.applicationEventCodeDuringEnrolmentAndCompleted(
+        COEStatus.completed,
+      );
+    }).toThrow("Unexpected coe status.");
```
Comment (andrewsignori-aot) [inline sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/utils-service/_tests_/unit/application-event-code-during-enrolment-and-completed-utils.service.applicationEventCodeDuringEnrolmentAndCompleted.spec.ts:44]: IMO acronyms like COE, NOA, MSFAA, and all should always be uppercase along comments.

File: sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/ier12-file-detail.ts [hunk 6]
```diff
@@ -295,8 +295,8 @@ export class IER12FileDetail implements IER12FileLine {
     record.appendOptionalNumberWithFiller(this.mssMiscCostsAllowance, 10);
     record.appendOptionalNumberWithFiller(this.mssTransportCostsAllowance, 10);
     record.appendOptionalNumberWithFiller(this.mssExtraTransportCosts, 10);
-    record.appendOptionalStringWithFiller(this.applicationEventCode, 4);
-    record.appendOptionalFormattedDate(this.applicationEventDate);
+    record.appendStringWithFiller(this.applicationEventCode, 4);
+    record.appendFormattedDate(this.applicationEventDate);
```
Comment (dheepak-aot) [inline sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/ier12-file-detail.ts:299]: 👍

File: sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/utils-service/_tests_/unit/application-event-code-during-enrolment-and-completed-utils.service.applicationEventCodeDuringEnrolmentAndCompleted.spec.ts [hunk 7]
```diff
@@ -0,0 +1,46 @@
+import { COEStatus } from "@sims/sims-db";
```
Comment (dheepak-aot) [inline sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/utils-service/_tests_/unit/application-event-code-during-enrolment-and-completed-utils.service.applicationEventCodeDuringEnrolmentAndCompleted.spec.ts:1]: Is it possible  to reduce the length of file name? `application-event-code-during-enrolment-and-completed-utils.service.applicationEventCodeDuringEnrolmentAndCompleted.spec.ts`

In windows when I switch between branches it is not letting me to delete the file

image

File: sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/ier12-file-detail.ts (hunk 6 above)
Comment (guru-aot) [reply sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/ier12-file-detail.ts:299]: 👍

File: sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/utils-service/_tests_/unit/application-event-code-during-enrolment-and-completed-utils.service.applicationEventCodeDuringCompleted.spec.ts [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+  it(
+    `Should return ${ApplicationEventCode.COEA} when the disbursement schedule status is ${DisbursementScheduleStatus.Pending}, ` +
+      `COE status is ${COEStatus.completed}, disbursement date is same or after the cutoff date(i.e disbursement date + ${DISBURSEMENT_FILE_GENERATION_ANTICIPATION_DAYS}) ` +
+      `and does not have any ${RestrictionActionType.StopFullTimeDisbursement}.`,
+    async () => {
+      // Arrange and act.
+      const applicationEventCode =
+        await applicationEventCodeDuringEnrolmentAndCompletedUtilsService.applicationEventCodeDuringCompleted(
+          payload,
+          [[RestrictionActionType.StopFullTimeApply]],
```
Comment (dheepak-aot) [inline sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/utils-service/_tests_/unit/application-event-code-during-enrolment-and-completed-utils.service.applicationEventCodeDuringCompleted.spec.ts:101]: is it not `RestrictionActionType.StopFullTimeDisbursement` ? comment also says the same

**Comment by guru-aot** [comment]
Done with the review, please have a look on the other dev comments

File: sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/utils-service/_tests_/unit/application-event-code-during-enrolment-and-completed-utils.service.applicationEventCodeDuringCompleted.spec.ts [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+    async () => {
+      // Arrange and act.
+      const applicationEventCode =
+        await applicationEventCodeDuringEnrolmentAndCompletedUtilsService.applicationEventCodeDuringCompleted(
+          {
+            ...payload,
+            disbursementScheduleStatus: DisbursementScheduleStatus.Sent,
+            disbursementFeedbackErrors: [
+              {
+                errorCode: FULL_TIME_DISBURSEMENT_FEEDBACK_ERRORS[0],
```
Comment (dheepak-aot) [inline sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/utils-service/_tests_/unit/application-event-code-during-enrolment-and-completed-utils.service.applicationEventCodeDuringCompleted.spec.ts:154]: As a suggestion, I would say to declare the error code for arrange as const and use.

```ts
const [feedbackError] = FULL_TIME_DISBURSEMENT_FEEDBACK_ERRORS;
```

File: sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/utils-service/_tests_/unit/application-event-code-during-enrolment-and-completed-utils.service.applicationEventCodeDuringCompleted.spec.ts [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+
+      // Arrange and act.
+      const applicationEventCode =
+        await applicationEventCodeDuringEnrolmentAndCompletedUtilsService.applicationEventCodeDuringCompleted(
+          {
+            ...payload,
+            disbursementScheduleStatus: DisbursementScheduleStatus.Sent,
+            disbursementFeedbackErrors: [
+              {
+                errorCode: "XXXXX",
```
Comment (dheepak-aot) [inline sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/utils-service/_tests_/unit/application-event-code-during-enrolment-and-completed-utils.service.applicationEventCodeDuringCompleted.spec.ts:182]: 👍

File: sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/utils-service/_tests_/unit/application-event-code-utils.service.getApplicationEventCode.spec.ts [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+    activeRestrictionsActionTypes = [];
+  });
+
+  it(`Should return ${ApplicationEventCode.DISC} when the application status is ${ApplicationStatus.Cancelled}.`, async () => {
+    // Act
+    const applicationEventCode =
+      await applicationEventCodeUtilsService.getApplicationEventCode(
+        applicationNumber,
+        ApplicationStatus.Cancelled,
+        payload,
```
Comment (dheepak-aot) [inline sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/utils-service/_tests_/unit/application-event-code-utils.service.getApplicationEventCode.spec.ts:54]: Suggestion I would say we can change the COEStatus to Pending when application status is cancelled to be inline with with more real time data.

File: sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/utils-service/_tests_/unit/application-event-code-utils.service.getApplicationEventCode.spec.ts [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+
+  it(`Should call applicationEventCodeDuringAssessment when the application status is ${ApplicationStatus.Assessment}.`, async () => {
+    // Act
+    await applicationEventCodeUtilsService.getApplicationEventCode(
+      applicationNumber,
+      ApplicationStatus.Assessment,
+      payload,
+    );
+    // Assert
+    expect(
```
Comment (dheepak-aot) [inline sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/utils-service/_tests_/unit/application-event-code-utils.service.getApplicationEventCode.spec.ts:68]: To clarify, isn't our intention here to assert the final event code obtained? I know we don't have to cover all the scenrios which happens inside this method as they are covered in the unit tests for child methods.

**Comment by dheepak-aot** [comment]
Nice work @ann-aot . Please have a look at the comments.

I totally missed something in the past PR about `hasAwardWithheldDueToRestriction`. We could have used the disbursement values returned from the getPendingAssessment query. I am not going to ask for any change but to record the point. Please let me know if I am missing something here.

File: sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/utils-service/_tests_/unit/application-event-code-during-enrolment-and-completed-utils.service.applicationEventCodeDuringCompleted.spec.ts (hunk 8 above)
Comment (ann-aot) [reply sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/utils-service/_tests_/unit/application-event-code-during-enrolment-and-completed-utils.service.applicationEventCodeDuringCompleted.spec.ts:101]: Its the false case (i.e `does not have any ${RestrictionActionType.StopFullTimeDisbursement}`)
image

**Review by andrewsignori-aot — COMMENTED** [review]
Great work with the unit test. Please take a look at the minor comments.

File: sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/utils-service/_tests_/unit/application-event-code-during-enrolment-and-completed-utils.service.applicationEventCodeDuringEnrolmentAndCompleted.spec.ts (hunk 7 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/utils-service/_tests_/unit/application-event-code-during-enrolment-and-completed-utils.service.applicationEventCodeDuringEnrolmentAndCompleted.spec.ts:1]: @dheepak-aot did the command `git config --system core.longpaths true` worked for you?

**Review by andrewsignori-aot — APPROVED** [review]
Thanks for making the changes and for taking care of the change suggested for `hasAwardWithheldDueToRestriction`. Looks good 👍 
Thanks @dheepak-aot for the nice catch 😉

File: sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/utils-service/_tests_/unit/application-event-code-during-enrolment-and-completed-utils.service.applicationEventCodeDuringEnrolmentAndCompleted.spec.ts (hunk 7 above)
Comment (dheepak-aot) [reply sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/utils-service/_tests_/unit/application-event-code-during-enrolment-and-completed-utils.service.applicationEventCodeDuringEnrolmentAndCompleted.spec.ts:1]: Yes it worked for me. Updating the windows settings alone did not work.

File: sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/utils-service/application-event-code-during-enrolment-and-completed-utils.service.ts [hunk 13]
```diff
@@ -60,30 +56,50 @@ export class ApplicationEventCodeDuringEnrolmentAndCompletedUtilsService {
...
   }
 
+  /**
+   * Checks if there is any partial or full award amount, that
+   * was withheld due to a restriction during an e-cert generation.
+   * @param disbursementValues disbursement values.
+   * @returns true if there is any partial or full award amount, that
+   * was withheld due to a restriction.
+   */
+  private hasAwardWithheldDueToRestriction(
```
Comment (dheepak-aot) [inline sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/utils-service/application-event-code-during-enrolment-and-completed-utils.service.ts:74]: Thanks for making this change 👍

**Review by dheepak-aot — APPROVED** [review]
Great job! thanks for doing the withheld restriction logic change in last minute.

**Review by guru-aot — APPROVED** [review]
LGTM, @ann-aot nice work

**Review by sh16011993 — APPROVED** [review]
Nice work @ann-aot 👍

File: sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/utils-service/_tests_/unit/application-event-date-utils.service.getApplicationEventDate.spec.ts [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+        payloadDisbursementSchedule,
+      );
+
+    // Assert
+    const [{ updatedAt }] =
+      payloadDisbursementSchedule.disbursementFeedbackErrors;
+    expect(applicationEventDate).toBe(updatedAt);
+  });
+
+  it(`Should return payloadDisbursementSchedule.disbursementDate - DISBURSEMENT_FILE_GENERATION_ANTICIPATION_DAYS when application event code is ${ApplicationEventCode.DISR}.`, () => {
```
Comment (andrepestana-aot) [inline sources/packages/backend/libs/integrations/src/institution-integration/ier12-integration/utils-service/_tests_/unit/application-event-date-utils.service.getApplicationEventDate.spec.ts:102]: Should `DISBURSEMENT_FILE_GENERATION_ANTICIPATION_DAYS` be evaluated here as the other vars or just a text like "disbursement file generation anticipation days"?

**Review by andrepestana-aot — APPROVED** [review]
Good job. Just left a minor comment.

**Pull request merged**